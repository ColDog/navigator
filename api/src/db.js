const config = require("./config");
const db = require("knex")(config.database[config.environment]);

module.exports = db;
