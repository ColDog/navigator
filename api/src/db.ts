import * as config from "./config";
import * as knex from "knex";
import * as log from './log';

log.info("db connection", config.database[config.environment].client);

export default knex(config.database[config.environment]);
