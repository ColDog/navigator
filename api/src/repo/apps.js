const validate = require("jsonschema").validate;
const db = require("../db");
const { NotFoundError, ValidationError } = require("../errors");

const schema = {
  $schema: "http://json-schema.org/draft-07/schema#",
  description: "Application Schema",
  type: "object",
  required: ["stages", "name"],
  properties: {
    name: { type: "string" },
    stages: {
      type: "array",
      items: {
        type: "object",
        required: ["name", "clusters"],
        properties: {
          name: { type: "string" },
          review: { type: "boolean" },
          promote: { type: "boolean" },
          auto: { type: "boolean" },
          clusters: {
            type: "array",
            items: {
              type: "object",
              required: ["name"],
              properties: {
                name: { type: "string" },
                values: { type: "object" }
              }
            }
          }
        }
      }
    }
  }
};

async function list() {
  return await db.select("name").from("apps");
}

async function fetch(name) {
  const app = await db
    .select("*")
    .from("apps")
    .where("name", name)
    .first();
  if (!app) {
    throw new NotFoundError("App not found");
  }
  return {
    name: app.name,
    stages: JSON.parse(app.stages)
  };
}

async function insert(app) {
  const result = validate(app || {}, schema);
  if (result.errors.length > 0) {
    throw new ValidationError("App is invalid", result.errors);
  }

  await db.transaction(async tx => {
    const row = await tx
      .select("name")
      .from("apps")
      .where("name", app.name);
    if (row.length > 0) {
      await tx.table("apps").update({ stages: JSON.stringify(app.stages) });
    } else {
      await tx
        .table("apps")
        .insert({ stages: JSON.stringify(app.stages), name: app.name });
    }
  });
}

module.exports = {
  list,
  insert,
  fetch
};
