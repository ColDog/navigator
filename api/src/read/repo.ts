import * as knex from "knex";
import db from "../db";
import { NotFoundError } from "../errors";

export class QuerySet<T> {
  name: string;
  load: (data: any) => T;

  constructor(name: string, load: (data: any) => T) {
    this.load = load;
    this.name = name;
  }

  public async find(cb: (db: knex) => knex.QueryBuilder): Promise<T> {
    const data = await cb(db);
    if (!data) {
      throw new NotFoundError(`${this.name} Does not exist`);
    }
    return this.load(data);
  }

  public async query(cb: (db: knex) => knex.QueryBuilder): Promise<T[]> {
    return await cb(db).map(this.load);
  }

  public async exists(cb: (db: knex) => knex.QueryBuilder): Promise<boolean> {
    const data = await cb(db);
    return !!data;
  }

  public table(name: string) {
    return db.table(name);
  }

  public db() {
    return db
  }
}
