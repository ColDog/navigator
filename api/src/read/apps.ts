import { QuerySet } from "./repo";
import { Upserted } from "../write/apps";
import { subscribe } from "../write";
import * as Knex from "knex";

const db = new QuerySet(
  "App",
  (data: any): App => ({
    ...data,
    stages: JSON.parse(data.stages)
  })
);

export interface Cluster {
  name: string;
  namespace: string;
  values: object;
}

export interface Stage {
  name: string;
  promote?: boolean;
  auto?: boolean;
  review?: boolean;
  clusters: Cluster[];
}

export interface App {
  id: string;
  name: string;
  chart?: string;
  deploy?: string;
  stages: Stage[];
  modified: string;
  created: string;
}

export async function list(): Promise<Array<{ id: string; name: string }>> {
  return await db.table("apps").select(["id", "name"]);
}

export async function get(name: string): Promise<App> {
  return await db.find(t =>
    t
      .select("*")
      .from("apps")
      .where("name", name)
      .first()
  );
}

async function upsert(tx: Knex.Transaction, app: Upserted) {
  const row = await tx
    .select("name")
    .from("apps")
    .where("name", app.name);
  if (row.length > 0) {
    await tx.table("apps").update({
      modified: new Date().toISOString(),
      chart: app.chart,
      deploy: app.deploy,
      stages: JSON.stringify(app.stages)
    });
  } else {
    await tx.table("apps").insert({
      stages: JSON.stringify(app.stages),
      chart: app.chart,
      name: app.name,
      deploy: app.deploy,
      modified: new Date().toISOString(),
      created: new Date().toISOString()
    });
  }
}

subscribe("apps.upserted", upsert);
