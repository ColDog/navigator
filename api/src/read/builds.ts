import { QuerySet } from "./repo";
import { subscribe } from "../write";
import * as Knex from "knex";
import { Created } from "../write/builds";

const db = new QuerySet<Build>({
  name: "Build",
  created: true,
  serialize: ["values"]
});

export interface Build {
  id: string;
  app: string;
  stage: string;
  version: string;
  chart?: string;
  namespace?: string;
  values?: object;
  created: string;
}

export async function list(app: string): Promise<Build[]> {
  return await db.query(t =>
    t
      .select("*")
      .from("builds")
      .where("app", app)
  );
}

export async function listUnreleased(
  app: string,
  stage: string
): Promise<Build[]> {
  return await db.query(t =>
    t
      .select("builds.*")
      .from("builds")
      .where("builds.app", app)
      .where("builds.stage", stage)
      .whereNull("releases.id")
      .leftJoin("releases", function() {
        this.on("builds.version", "=", "releases.version")
          .andOn("builds.stage", "=", "releases.stage")
          .andOn("builds.app", "=", "releases.app");
      })
  );
}

export async function exists(app: string, stage: string, version: string) {
  return await db.exists(t =>
    t
      .select("id")
      .from("builds")
      .where({ app, stage, version })
      .first()
  );
}

export async function get(app: string, stage: string, version: string) {
  return await db.find(t =>
    t
      .select("*")
      .from("builds")
      .where({ app, stage, version })
      .first()
  );
}

export async function latest(
  app: string,
  stage: string
): Promise<Build | undefined> {
  const bs = await db.query(t =>
    t
      .select("*")
      .from("builds")
      .where({ app, stage })
      .orderBy("id", "desc")
  );
  return bs[0];
}

export async function lastByNamespace(
  app: string,
  stage: string,
  n: number = 25
): Promise<Build[]> {
  const builds = await last(app, stage, n);
  const namespaces = builds
    .map(b => b.namespace)
    .filter((ns, i, s) => s.indexOf(ns) === i); // Unique namespaces.
  const byNamespace = namespaces
    .map(ns => builds.find(b => b.namespace === ns))
    .filter(b => b);
  return byNamespace as Build[];
}

export async function last(app: string, stage: string, n: number = 25) {
  return await db.query(t =>
    t
      .select("*")
      .from("builds")
      .where({ app, stage })
      .orderBy("id", "desc")
      .limit(n)
  );
}

async function insert(tx: Knex.Transaction, build: Created) {
  await db.create(tx.table("builds"), build);
}

subscribe("builds.created", insert);
