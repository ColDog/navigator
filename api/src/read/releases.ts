import { QuerySet } from "./repo";
import { subscribe } from "../write";
import * as Knex from "knex";
import { Created, Updated, Status } from "../write/releases";
import { NotFoundError } from "../errors";

const db = new QuerySet<Release>({
  name: "Release",
  created: true,
  modified: true,
  serialize: ["canary"],
  booleans: ["removal"]
});

export interface Release {
  id: string;
  app: string;
  stage: string;
  version: string;
  canary?: {
    weight: number;
    version: string;
  };
  removal: boolean;
  worker?: string;
  status?: string;
  modified: string;
  created: string;
}

export async function get(id: string): Promise<Release> {
  return await db.find(t =>
    t
      .select("*")
      .from("releases")
      .where({ id })
      .first()
  );
}

export async function listByStage(
  app: string,
  stage: string,
  n: number = 25
): Promise<Release[]> {
  return await db.query(t =>
    t
      .select("*")
      .from("releases")
      .where({ app, stage })
      .orderBy("id", "DESC")
      .limit(n)
  );
}

export async function getByApp(
  app: string,
  stage: string,
  version: string
): Promise<Release> {
  return await db.find(t =>
    t
      .select("*")
      .from("releases")
      .where({ app, stage, version })
      .orderBy("id", "DESC")
      .first()
  );
}

export async function previous(app: string, stage: string): Promise<Release> {
  const list = await listByStage(app, stage, 2);
  if (list.length !== 2) {
    throw new NotFoundError(`Previous release not found ${app}/${stage}`);
  }
  return list[1]; // Preceding release.
}

export async function pop(worker: string) {
  let releaseId;
  await db.db().transaction(async tx => {
    const row = await tx
      .select("id")
      .from("releases")
      .where("worker", null)
      .first();
    if (row) {
      releaseId = row.id;
      await tx
        .table("releases")
        .update({ worker })
        .where("id", releaseId);
    }
  });
  return releaseId;
}

async function insert(tx: Knex.Transaction, release: Created) {
  await db.create(tx.table("releases"), release);
}

async function update(tx: Knex.Transaction, release: Updated) {
  await db.update(tx.table("releases").where("id", release.id), {
    status: release.status
  });
}

async function invalid(tx: Knex.Transaction, payload: { releaseId: string }) {
  await db.update(tx.table("releases").where("id", payload.releaseId), {
    status: Status.Invalid
  });
}

subscribe("releases.created", insert);
subscribe("releases.updated", update);
subscribe("releases.invalid", invalid);
