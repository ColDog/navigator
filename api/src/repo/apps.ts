import { validate } from "jsonschema";
import { ValidationError } from "../errors";
import { QuerySet } from "./repo";
import * as crypto from "crypto";
import _ = require("lodash");

const db = new QuerySet(
  (data: any): App => ({
    ...data,
    stages: JSON.parse(data.stages)
  })
);

export interface Cluster {
  name: string;
  namespace: string;
  values: object;
}

export interface Stage {
  name: string;
  promote?: boolean;
  auto?: boolean;
  review?: boolean;
  clusters: Cluster[];
}

export interface App {
  id: string;
  name: string;
  chart?: string;
  deploy?: string;
  stages: Stage[];
  modified: string;
  created: string;
}

const schema = {
  $schema: "http://json-schema.org/draft-07/schema#",
  description: "Application Schema",
  type: "object",
  required: ["stages", "name"],
  properties: {
    name: { type: "string" },
    chart: { type: "string" },
    stages: {
      type: "array",
      items: {
        type: "object",
        required: ["name", "clusters"],
        properties: {
          name: { type: "string" },
          review: { type: "boolean" },
          promote: { type: "boolean" },
          auto: { type: "boolean" },
          clusters: {
            type: "array",
            items: {
              type: "object",
              required: ["name", "namespace"],
              properties: {
                namespace: { type: "string" },
                name: { type: "string" },
                values: { type: "object" }
              }
            }
          }
        }
      }
    }
  }
};

export async function list(): Promise<Array<{ id: string; name: string }>> {
  return await db.table("apps").select(["id", "name"]);
}

export async function get(name: string): Promise<App> {
  return await db.find(t =>
    t
      .select("*")
      .from("apps")
      .where("name", name)
      .first()
  );
}

export async function getKey(name: string): Promise<string> {
  const app = await db
    .table("apps")
    .select({ key: "modified" })
    .where("name", name)
    .first();
  const release = await db
    .table("releases")
    .max({ key: "modified" })
    .where("app", name)
    .first();
  const build = await db
    .table("builds")
    .max({ key: "id" })
    .where("app", name)
    .first();

  const md5 = crypto.createHash("md5");
  md5.update(`${_.get(build, "key", "-")}`);
  md5.update(`${_.get(app, "key", "-")}`);
  md5.update(`${_.get(release, "key", "-")}`);
  return md5.digest("hex");
}

export async function insert(app: any) {
  const result = validate(app || {}, schema, { allowUnknownAttributes: false });
  if (result.errors.length > 0) {
    throw new ValidationError("App is invalid", result.errors);
  }

  await db.db().transaction(async tx => {
    const row = await tx
      .select("name")
      .from("apps")
      .where("name", app.name);
    if (row.length > 0) {
      await tx.table("apps").update({
        modified: new Date().toISOString(),
        chart: app.chart,
        stages: JSON.stringify(app.stages)
      });
    } else {
      await tx.table("apps").insert({
        stages: JSON.stringify(app.stages),
        chart: app.chart,
        name: app.name,
        modified: new Date().toISOString(),
        created: new Date().toISOString()
      });
    }
  });
}
