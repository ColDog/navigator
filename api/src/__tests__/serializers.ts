import * as builds from "../repo/builds";
import * as apps from "../repo/apps";
import * as releases from "../repo/releases";
import * as MockDate from "mockdate";
import {
  serializeBuild,
  serializeRelease,
  appSerializer,
} from "../serializers";
import db from "../db";

beforeEach(async () => {
  await db.migrate.latest();
});

const u = { email: "test" };

beforeEach(async () => {
  MockDate.set(new Date(2013, 9, 23));
});

describe("serializers", () => {
  it("serializes a build", async () => {
    await builds.insert(u, { app: "test", version: "v1", stage: "review" });
    const build = await builds.get("test", "review", "v1");
    expect(await serializeBuild(build)).toMatchSnapshot();
  });

  it("serializes a release", async () => {
    await builds.insert(u, { app: "test", version: "v2", stage: "review" });
    await releases.insert(u, { app: "test", version: "v2", stage: "review" });

    const release = await releases.getByApp("test", "review", "v2");
    expect(await serializeRelease(release)).toMatchSnapshot();
  });

  it("serializes an app", async () => {
    await apps.insert(u, {
      name: "test",
      config: {
        chart: "service",
        deploy: "mock-deploy",
      },
      stages: [
        { name: "review", clusters: [{ name: "test", namespace: "default" }] },
      ],
    });
    await builds.insert(u, { app: "test", version: "v3", stage: "review" });
    await releases.insert(u, { app: "test", version: "v3", stage: "review" });

    const app = await apps.get("test");
    expect(await appSerializer(app)).toMatchSnapshot();
  });
});
