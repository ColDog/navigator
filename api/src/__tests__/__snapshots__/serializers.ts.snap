// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`serializers serializes a build 1`] = `
Object {
  "app": "test",
  "created": "2013-10-23T07:00:00.000Z",
  "id": 1,
  "namespace": null,
  "releaseId": null,
  "released": false,
  "removed": false,
  "stage": "review",
  "status": "UNRELEASED",
  "values": Object {},
  "version": "v1",
}
`;

exports[`serializers serializes a release 1`] = `
Object {
  "app": "test",
  "created": "2013-10-23T07:00:00.000Z",
  "id": 2,
  "namespace": null,
  "releaseId": 1,
  "released": true,
  "removed": false,
  "stage": "review",
  "status": "PENDING",
  "values": Object {},
  "version": "v2",
}
`;

exports[`serializers serializes an app 1`] = `
Object {
  "chart": "service",
  "created": "2013-10-23T07:00:00.000Z",
  "deploy": "./deployer-mock.sh",
  "id": 1,
  "manifest": Object {
    "chart": "service",
    "deploy": "./deployer-mock.sh",
    "name": "test",
    "stages": Array [
      Object {
        "clusters": Array [
          Object {
            "name": "test",
            "namespace": "default",
          },
        ],
        "name": "review",
      },
    ],
  },
  "modified": "2013-10-23T07:00:00.000Z",
  "name": "test",
  "stages": Array [
    Object {
      "builds": Array [
        Object {
          "app": "test",
          "created": "2013-10-23T07:00:00.000Z",
          "id": 3,
          "namespace": null,
          "releaseId": 2,
          "released": true,
          "removed": false,
          "stage": "review",
          "status": "PENDING",
          "values": Object {},
          "version": "v3",
        },
      ],
      "clusters": Array [
        Object {
          "name": "test",
          "namespace": "default",
        },
      ],
      "current": Object {
        "app": "test",
        "created": "2013-10-23T07:00:00.000Z",
        "id": 3,
        "namespace": null,
        "releaseId": 2,
        "released": true,
        "removed": false,
        "stage": "review",
        "status": "PENDING",
        "values": Object {},
        "version": "v3",
      },
      "name": "review",
      "previous": Object {
        "app": "test",
        "created": "2013-10-23T07:00:00.000Z",
        "id": 2,
        "namespace": null,
        "releaseId": 1,
        "released": true,
        "removed": false,
        "stage": "review",
        "status": "PENDING",
        "values": Object {},
        "version": "v2",
      },
      "released": Object {
        "app": "test",
        "created": "2013-10-23T07:00:00.000Z",
        "id": 3,
        "namespace": null,
        "releaseId": 2,
        "released": true,
        "removed": false,
        "stage": "review",
        "status": "PENDING",
        "values": Object {},
        "version": "v3",
      },
    },
  ],
}
`;
